hw/drivers/ram/spiram/src/spiram.c

```diff
@@ -0,0 +1,286 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include <string.h>
+#include <assert.h>
+
+#include "os/mynewt.h"
+#include <bsp/bsp.h>
+#include <hal/hal_spi.h>
+#include <hal/hal_gpio.h>
+#include <spiram/spiram.h>
+#if MYNEWT_VAL(BUS_DRIVER_PRESENT)
+#include <bus/drivers/spi_common.h>
+#endif
+
+static inline void
+spiram_lock(struct spiram_dev *dev)
+{
+#if MYNEWT_VAL(OS_SCHEDULING)
+    os_mutex_pend(&dev->lock, OS_TIMEOUT_NEVER);
+#endif
+}
+
+static inline void
+spiram_unlock(struct spiram_dev *dev)
+{
+#if MYNEWT_VAL(OS_SCHEDULING)
+    os_mutex_release(&dev->lock);
+#endif
+}
+
+void
+spiram_cs_activate(struct spiram_dev *dev)
+{
+#if MYNEWT_VAL(BUS_DRIVER_PRESENT)
+    struct bus_spi_node *node = (struct bus_spi_node *)&dev->dev;
+    hal_gpio_write(node->pin_cs, 0);
+#else
+    hal_gpio_write(dev->ss_pin, 0);
+#endif
+}
+
+void
+spiram_cs_deactivate(struct spiram_dev *dev)
+{
+#if MYNEWT_VAL(BUS_DRIVER_PRESENT)
+    struct bus_spi_node *node = &dev->dev;
+    hal_gpio_write(node->pin_cs, 1);
+#else
+    hal_gpio_write(dev->ss_pin, 1);
+#endif
+}
+
+int
+spiram_write_enable(struct spiram_dev *dev)
+{
+    int rc = 0;
+    spiram_lock(dev);
+
+    if (dev->characteristics->write_enable_cmd) {
+#if MYNEWT_VAL(BUS_DRIVER_PRESENT)
+        rc = bus_node_simple_write((struct os_dev *)&dev->dev,
+                              &dev->characteristics->write_enable_cmd, 1);
+#else
+        spiram_cs_activate(dev);
+
+        rc = hal_spi_tx_val(dev->spi_num, dev->characteristics->write_enable_cmd);
+
+        spiram_cs_deactivate(dev);
+#endif
+    }
+
+    spiram_unlock(dev);
+
+    return rc;
+}
+
+int
+spiram_read(struct spiram_dev *dev, uint32_t addr, void *buf, uint32_t size)
+{
+    uint8_t cmd[5] = { SPIRAM_READ };
+    int i;
+    int rc = 0;
+    int cmd_size = 1 + dev->characteristics->address_bytes + dev->characteristics->dummy_bytes;
+
+    i = dev->characteristics->address_bytes;
+    while (i) {
+        cmd[i] = (uint8_t)addr;
+        addr >>= 8;
+        --i;
+    }
+    spiram_lock(dev);
+
+    if (size > 0) {
+#if MYNEWT_VAL(BUS_DRIVER_PRESENT)
+        rc = bus_node_simple_write_read_transact((struct os_dev *)&dev->dev,
+                                                 &cmd, cmd_size, buf, size);
+#else
+        spiram_cs_activate(dev);
+
+        /* Send command + address */
+        rc = hal_spi_txrx(dev->spi_num, cmd, NULL, cmd_size);
+        if (rc == 0) {
+            /* For security mostly, do not output random data, fill it with FF */
+            memset(buf, 0xFF, size);
+            /* Tx buf does not matter, for simplicity pass read buffer */
+            rc = hal_spi_txrx(dev->spi_num, buf, buf, size);
+        }
+
+        spiram_cs_deactivate(dev);
+#endif
+    }
+
+    spiram_unlock(dev);
+
+    return rc;
+}
+
+int
+spiram_write(struct spiram_dev *dev, uint32_t addr, void *buf, uint32_t size)
+{
+    uint8_t cmd[4] = { SPIRAM_WRITE };
+    int rc = 0;
+    int i;
```

Comment:
Why is this an int ? A uint8_t is enough.

Reply:
see above